pingcap/tiflash
# Support rate limit for background tasks of StorageDeltaMerge

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tics/issues/1194

Problem Summary: Throughput on ebs disk is limited(i.e. 250 MB/s), we need rate limit write rate of background tasks to decrease the interference with other tasks.

### What is changed and how it works?

Proposal: https://docs.google.com/document/d/1fNKbF6hoo1uErylvjOlQmkNRJXeY6-DtNLe-pFCcoeY/edit

What's Changed: Add a RateLimiter class which manages write rate of background tasks. Currently, tasks including Segment Merge, MergeDelta and Physical Split will be rate limited by this class.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Manual test
Use tidb-lightning(use tidb-backend mode) to load tpcc(warehouse=1000) data to tidb and check the write flow metrics of tiflash. From the metics, we can see that the io flow is being rate limited as expected.
![image](https://user-images.githubusercontent.com/47731263/101602627-62cbc380-3a39-11eb-8fae-72de8d0f3dd3.png)

Side effects

- Performance regression
    - Consumes more CPU

### Release note 

- Support io rate limit of background tasks of the storage engine.
